### PR TITLE
Add Joint Bilateral Filter

### LINKED ISSUE #2244: Add Joint Bilateral Filter
## 🚀 Feature

Joint Bilateral Filter as described in the paper http://vis.berkeley.edu/papers/fnf/fnfSmall.pdf

Compared to Bilateral Filter, the only difference is using another image (aka guidance) to calculate the color/intensity kernel.

OpenCV: https://docs.opencv.org/4.x/da/d17/group__ximgproc__filters.html#ga80b9b58fb85dd069691b709285ab985c

Once #2242 is merged, extending the current code for Joint Bilateral Filter should be straight-forward.

## Additional context

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by gau-nernst**
#### Changes

Closes #2244

TODO

- [x] Add docs
- [x] Add tests
- [x] Add test against OpenCV
- [x] Check forward signature for Module

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by gau-nernst** [comment]
Sanity check by visual comparison against OpenCV. Observe the edge of the pepper has been "transferred" to Lena's hair.

jbf

```python
import io

import cv2
import matplotlib.pyplot as plt
import numpy as np
import requests
import torch
from PIL import Image

from kornia.filters import joint_bilateral_blur
from kornia.filters.bilateral import joint_bilateral_blur


def load_img(url):
    img =  np.array(Image.open(io.BytesIO(requests.get(url).content)))
    img = img.astype(np.float32) / 255
    return img

img = load_img("https://upload.wikimedia.org/wikipedia/en/7/7d/Lenna_%28test_image%29.png")
guidance = load_img("https://sipi.usc.edu/database/download.php?vol=misc&img=4.2.07")

img_torch = torch.from_numpy(img).permute(2, 0, 1).unsqueeze(0)
guidance_torch = torch.from_numpy(guidance).permute(2, 0, 1).unsqueeze(0)

k = 11
c = 0.1
s = 2

out1 = joint_bilateral_blur(img_torch, guidance_torch, k, c, (s, s))
out1 = out1.squeeze(0).permute(1, 2, 0).numpy()
out2 = cv2.ximgproc.jointBilateralFilter(guidance, img, k, c, s)

diff = np.abs(out1 - out2) / out2
print(diff.mean(), diff.max(), diff.min())  # 0.0016992172 0.307278 0.0

fig, axes = plt.subplots(2, 3, figsize=(15, 10))
fig.delaxes(axes[0][2])

axes[0][0].imshow((img * 255).astype(np.uint8))
axes[0][0].set_title("Image")

axes[0][1].imshow((guidance * 255).astype(np.uint8))
axes[0][1].set_title("Guidance")

axes[1][0].imshow((out1 * 255).astype(np.uint8))
axes[1][0].set_title("Pytorch")

axes[1][1].imshow((out2 * 255).astype(np.uint8))
axes[1][1].set_title("OpenCV")

axes[1][2].imshow((np.abs(out1 - out2) * 255).astype(np.uint8))
axes[1][2].set_title("Difference")

plt.tight_layout()
plt.savefig("jbf.png")
```

File: kornia/filters/bilateral.py [hunk 1]
```diff
@@ -1,13 +1,64 @@
...
+    guidance: Tensor | None,
+    kernel_size: tuple[int, int] | int,
+    sigma_color: float | Tensor,
+    sigma_space: tuple[float, float] | Tensor,
+    border_type: str = 'reflect',
+    color_distance_type: str = "l1",
+) -> Tensor:
+    "Single implementation for both Bilateral Filter and Joint Bilateral Filter"
+
+    KORNIA_CHECK_IS_TENSOR(input)
```
Comment (edgarriba) [inline kornia/filters/bilateral.py:21]: Is tensor is included in check shape

File: kornia/filters/bilateral.py [hunk 2]
```diff
@@ -1,13 +1,64 @@
...
+    KORNIA_CHECK_SHAPE(input, ['B', 'C', 'H', 'W'])
+    if guidance is not None:
+        KORNIA_CHECK_IS_TENSOR(guidance)
```
Comment (edgarriba) [inline kornia/filters/bilateral.py:24]: The three lines below could be reduced to input.shape==guidance.shape ?

File: kornia/filters/bilateral.py [hunk 3]
```diff
@@ -1,13 +1,64 @@
...
+    unfolded_input = padded_input.unfold(2, ky, 1).unfold(3, kx, 1).flatten(-2)  # (B, C, H, W, K x K)
+
+    if guidance is None:
+        guidance = input
+        unfolded_guidance = unfolded_input
+    else:
+        padded_guidance = pad(guidance, (pad_x, pad_x, pad_y, pad_y), mode=border_type)
+        unfolded_guidance = padded_guidance.unfold(2, ky, 1).unfold(3, kx, 1).flatten(-2)  # (B, C, H, W, K x K)
+
+    diff = unfolded_guidance - guidance.unsqueeze(-1)
```
Comment (edgarriba) [inline kornia/filters/bilateral.py:45]: One cool thing to explore in the future is that we can pass a lambda fun to customise the distance function

File: kornia/filters/bilateral.py [hunk 4]
```diff
@@ -48,71 +99,54 @@ def bilateral_blur(
...
+def joint_bilateral_blur(
+    input: Tensor,
+    guidance: Tensor,
+    kernel_size: tuple[int, int] | int,
+    sigma_color: float | Tensor,
+    sigma_space: tuple[float, float] | Tensor,
+    border_type: str = 'reflect',
+    color_distance_type: str = "l1",
+) -> Tensor:
+    r"""Blur a tensor using a Joint Bilateral filter.
```
Comment (edgarriba) [inline kornia/filters/bilateral.py:115]: Add the generated image.

**Review by edgarriba — APPROVED** [review]
LGTM

File: kornia/filters/bilateral.py (hunk 1 above)
Comment (gau-nernst) [reply kornia/filters/bilateral.py:21]: From what I see, currently `CHECK_SHAPE()` does not have `IS_TENSOR()` check

File: kornia/filters/bilateral.py (hunk 2 above)
Comment (gau-nernst) [reply kornia/filters/bilateral.py:24]: I want to allow guidance and input having different number of channels e.g. guidance is grayscale, input is RGB. What do you think? Should we allow that?

File: kornia/filters/bilateral.py (hunk 4 above)
Comment (gau-nernst) [reply kornia/filters/bilateral.py:115]: What should be the guidance image?

Comment (edgarriba) [reply kornia/filters/bilateral.py:115]: Anything related to Lena :)
http://www.lenna.org/

File: kornia/filters/bilateral.py (hunk 2 above)
Comment (edgarriba) [reply kornia/filters/bilateral.py:24]: I guess it will only work for guidance being num channels 1 or same as input, right ? 1 Because of Broadcasting

File: kornia/filters/bilateral.py (hunk 1 above)
Comment (edgarriba) [reply kornia/filters/bilateral.py:21]: Ok, that changed. Fair enough

File: kornia/filters/bilateral.py (hunk 2 above)
Comment (gau-nernst) [reply kornia/filters/bilateral.py:24]: `color_distance` will always have 1 channel since we doing reduction across channel dimension (i.e. `.sum(dim=1, keepdim=True)`). So it doesn't matter how many channels `guidance` has. The intensity/color kernel to be applied is the same across all channels of `input`.

Comment (edgarriba) [reply kornia/filters/bilateral.py:24]: i don't have any preference here -- up to you in terms of what you think that can be useful for other to research/experiment

File: kornia/filters/bilateral.py (hunk 4 above)
Comment (gau-nernst) [reply kornia/filters/bilateral.py:119]: I'm using the simba image from `kornia.color` sample image. What do you think?

image

Comment (johnnv1) [reply kornia/filters/bilateral.py:119]: I like this one, it looks easy to saw at least

**Review by johnnv1 — APPROVED** [review]

**Comment by gau-nernst** [comment]
@edgarriba Can I mark the remaining conversations as resolved? Do you want anything else changed?
